In [3]:
# Install numpy and pandas packages


import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [4]:
users= pd.read_fwf('zee-users.dat',encoding='ISO-8859-1')
users.head()

UserID::Gender::Age::Occupation::Zip-code
0                        1::F::1::10::48067
1                       2::M::56::16::70072
2                       3::M::25::15::55117
3                        4::M::45::7::02460
4                       5::M::25::20::55455

In [5]:
users=users['UserID::Gender::Age::Occupation::Zip-code'].str.split("::",expand=True)
users.columns=['UserId','Gender','Age','Occupation','Zip-code']
users

UserId Gender Age Occupation Zip-code
0         1      F   1         10    48067
1         2      M  56         16    70072
2         3      M  25         15    55117
3         4      M  45          7    02460
4         5      M  25         20    55455
...     ...    ...  ..        ...      ...
6035   6036      F  25         15    32603
6036   6037      F  45          1    76006
6037   6038      F  56          1    14706
6038   6039      F  45          0    01060
6039   6040      M  25          6    11106

[6040 rows x 5 columns]

In [6]:
movies= pd.read_fwf('zee-movies.dat',encoding='ISO-8859-1')
movies.head()

Movie ID::Title::Genres Unnamed: 1 Unnamed: 2
0  1::Toy Story (1995)::Animation|Children's|Comedy        NaN        NaN
1   2::Jumanji (1995)::Adventure|Children's|Fantasy        NaN        NaN
2        3::Grumpier Old Men (1995)::Comedy|Romance        NaN        NaN
3         4::Waiting to Exhale (1995)::Comedy|Drama        NaN        NaN
4     5::Father of the Bride Part II (1995)::Comedy        NaN        NaN

In [7]:
movies=movies['Movie ID::Title::Genres'].str.split("::",expand=True)
movies.columns=['MovieId','Title','Genres']
movies

MovieId                               Title                        Genres
0          1                    Toy Story (1995)   Animation|Children's|Comedy
1          2                      Jumanji (1995)  Adventure|Children's|Fantasy
2          3             Grumpier Old Men (1995)                Comedy|Romance
3          4            Waiting to Exhale (1995)                  Comedy|Drama
4          5  Father of the Bride Part II (1995)                        Comedy
...      ...                                 ...                           ...
3878    3948             Meet the Parents (2000)                        Comedy
3879    3949          Requiem for a Dream (2000)                         Drama
3880    3950                    Tigerland (2000)                         Drama
3881    3951             Two Family House (2000)                         Drama
3882    3952               Contender, The (2000)                Drama|Thriller

[3883 rows x 3 columns]

In [8]:
ratings= pd.read_fwf('zee-ratings.dat',encoding='ISO-8859-1')
ratings

UserID::MovieID::Rating::Timestamp
0                    1::1193::5::978300760
1                     1::661::3::978302109
2                     1::914::3::978301968
3                    1::3408::4::978300275
4                    1::2355::5::978824291
...                                    ...
1000204           6040::1091::1::956716541
1000205           6040::1094::5::956704887
1000206            6040::562::5::956704746
1000207           6040::1096::4::956715648
1000208           6040::1097::4::956715569

[1000209 rows x 1 columns]

In [9]:
ratings=ratings['UserID::MovieID::Rating::Timestamp'].str.split("::",expand=True)
ratings.columns=['UserId','MovieId','Rating','Timestamp']
ratings

UserId MovieId Rating  Timestamp
0            1    1193      5  978300760
1            1     661      3  978302109
2            1     914      3  978301968
3            1    3408      4  978300275
4            1    2355      5  978824291
...        ...     ...    ...        ...
1000204   6040    1091      1  956716541
1000205   6040    1094      5  956704887
1000206   6040     562      5  956704746
1000207   6040    1096      4  956715648
1000208   6040    1097      4  956715569

[1000209 rows x 4 columns]

In [10]:
df1= pd.merge(movies,ratings, how='inner',on='MovieId')
df1

MovieId                  Title                       Genres UserId  \
0             1       Toy Story (1995)  Animation|Children's|Comedy      1   
1             1       Toy Story (1995)  Animation|Children's|Comedy      6   
2             1       Toy Story (1995)  Animation|Children's|Comedy      8   
3             1       Toy Story (1995)  Animation|Children's|Comedy      9   
4             1       Toy Story (1995)  Animation|Children's|Comedy     10   
...         ...                    ...                          ...    ...   
1000204    3952  Contender, The (2000)               Drama|Thriller   5812   
1000205    3952  Contender, The (2000)               Drama|Thriller   5831   
1000206    3952  Contender, The (2000)               Drama|Thriller   5837   
1000207    3952  Contender, The (2000)               Drama|Thriller   5927   
1000208    3952  Contender, The (2000)               Drama|Thriller   5998   

        Rating   Timestamp  
0            5   978824268  
1            4   978237008  
2            4   978233496  
3            5   978225952  
4            5   978226474  
...        ...         ...  
1000204      4   992072099  
1000205      3   986223125  
1000206      4  1011902656  
1000207      1   979852537  
1000208      4  1001781044  

[1000209 rows x 6 columns]

In [11]:
data= pd.merge(df1,users, how='inner',on='UserId')
data

MovieId                  Title                       Genres UserId  \
0             1       Toy Story (1995)  Animation|Children's|Comedy      1   
1             1       Toy Story (1995)  Animation|Children's|Comedy      6   
2             1       Toy Story (1995)  Animation|Children's|Comedy      8   
3             1       Toy Story (1995)  Animation|Children's|Comedy      9   
4             1       Toy Story (1995)  Animation|Children's|Comedy     10   
...         ...                    ...                          ...    ...   
1000204    3952  Contender, The (2000)               Drama|Thriller   5812   
1000205    3952  Contender, The (2000)               Drama|Thriller   5831   
1000206    3952  Contender, The (2000)               Drama|Thriller   5837   
1000207    3952  Contender, The (2000)               Drama|Thriller   5927   
1000208    3952  Contender, The (2000)               Drama|Thriller   5998   

        Rating   Timestamp Gender Age Occupation Zip-code  
0            5   978824268      F   1         10    48067  
1            4   978237008      F  50          9    55117  
2            4   978233496      M  25         12    11413  
3            5   978225952      M  25         17    61614  
4            5   978226474      F  35          1    95370  
...        ...         ...    ...  ..        ...      ...  
1000204      4   992072099      F  25          7    92120  
1000205      3   986223125      M  25          1    92120  
1000206      4  1011902656      M  25          7    60607  
1000207      1   979852537      M  35         14    10003  
1000208      4  1001781044      M  18          4    61820  

[1000209 rows x 10 columns]

In [12]:
data.dtypes

MovieId       object
Title         object
Genres        object
UserId        object
Rating        object
Timestamp     object
Gender        object
Age           object
Occupation    object
Zip-code      object
dtype: object

In [13]:
# convert Rating to float
data['Rating']= data['Rating'].astype(float)

In [14]:
df= pd.pivot_table(data,index='UserId',columns='MovieId',values='Rating')
df

MovieId    1  10  100  1000  1002  1003  1004  1005  1006  1007  ...  99  990  \
UserId                                                           ...            
1        5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
10       5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
100      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1000     5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1001     4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
...      ...  ..  ...   ...   ...   ...   ...   ...   ...   ...  ...  ..  ...   
995      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
996      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
997      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
998      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
999      NaN NaN  NaN   NaN   NaN   2.0   3.0   NaN   NaN   NaN  ... NaN  NaN   

MovieId  991  992  993  994  996  997  998  999  
UserId                                           
1        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
10       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
100      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1000     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1001     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...      ...  ...  ...  ...  ...  ...  ...  ...  
995      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
996      NaN  NaN  NaN  NaN  NaN  NaN  NaN  3.0  
997      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
998      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
999      NaN  NaN  NaN  NaN  2.0  NaN  NaN  NaN  

[6040 rows x 3706 columns]

In [15]:
df.isna().sum()

MovieId
1       3963
10      5152
100     5912
1000    6020
1002    6032
        ... 
994     5590
996     5784
997     6012
998     5947
999     5754
Length: 3706, dtype: int64

In [16]:
# calculate sparsity , ratio of total number of zeros/null values and total values in the table
sparsity = df.isna().sum().sum()/df.size
print(sparsity)

0.9553163743776871


So , this is highly sparse data, 95.5% of the data is empty..

In [17]:
df

MovieId    1  10  100  1000  1002  1003  1004  1005  1006  1007  ...  99  990  \
UserId                                                           ...            
1        5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
10       5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
100      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1000     5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1001     4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
...      ...  ..  ...   ...   ...   ...   ...   ...   ...   ...  ...  ..  ...   
995      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
996      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
997      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
998      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
999      NaN NaN  NaN   NaN   NaN   2.0   3.0   NaN   NaN   NaN  ... NaN  NaN   

MovieId  991  992  993  994  996  997  998  999  
UserId                                           
1        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
10       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
100      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1000     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1001     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...      ...  ...  ...  ...  ...  ...  ...  ...  
995      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
996      NaN  NaN  NaN  NaN  NaN  NaN  NaN  3.0  
997      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
998      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
999      NaN  NaN  NaN  NaN  2.0  NaN  NaN  NaN  

[6040 rows x 3706 columns]

In [18]:
import numpy
print(numpy.__version__)


1.26.4


In [19]:

from surprise import SVD, Dataset, Reader
from surprise.model_selection import train_test_split
from surprise import accuracy

In [20]:
# import numpy
# print(numpy.__version__)


In [21]:
# import numpy
# import surprise

# print("NumPy Version:", numpy.__version__)
# print("Surprise Version:", surprise.__version__)


In [22]:
# pip uninstall -y numpy scikit-surprise


In [23]:
# pip install --no-cache-dir numpy==1.26.4


In [24]:
# pip install --no-cache-dir scikit-surprise


In [25]:
# from surprise import SVD, Dataset, Reader


In [26]:


# Step 2: Convert to Surprise Format (Ignoring NaNs Automatically)
reader = Reader(rating_scale=(1, 5))  # Rating scale 1-5
ratings_df = Dataset.load_from_df(data[['UserId','MovieId','Rating']], reader)  # Drop NaNs only here

# Step 3: Train-Test Split
trainset, testset = train_test_split(ratings_df, test_size=0.2, random_state=42)

# Step 4: Train SVD Model
model = SVD(n_factors=10, random_state=42)
model.fit(trainset)




In [27]:
data

MovieId                  Title                       Genres UserId  \
0             1       Toy Story (1995)  Animation|Children's|Comedy      1   
1             1       Toy Story (1995)  Animation|Children's|Comedy      6   
2             1       Toy Story (1995)  Animation|Children's|Comedy      8   
3             1       Toy Story (1995)  Animation|Children's|Comedy      9   
4             1       Toy Story (1995)  Animation|Children's|Comedy     10   
...         ...                    ...                          ...    ...   
1000204    3952  Contender, The (2000)               Drama|Thriller   5812   
1000205    3952  Contender, The (2000)               Drama|Thriller   5831   
1000206    3952  Contender, The (2000)               Drama|Thriller   5837   
1000207    3952  Contender, The (2000)               Drama|Thriller   5927   
1000208    3952  Contender, The (2000)               Drama|Thriller   5998   

         Rating   Timestamp Gender Age Occupation Zip-code  
0           5.0   978824268      F   1         10    48067  
1           4.0   978237008      F  50          9    55117  
2           4.0   978233496      M  25         12    11413  
3           5.0   978225952      M  25         17    61614  
4           5.0   978226474      F  35          1    95370  
...         ...         ...    ...  ..        ...      ...  
1000204     4.0   992072099      F  25          7    92120  
1000205     3.0   986223125      M  25          1    92120  
1000206     4.0  1011902656      M  25          7    60607  
1000207     1.0   979852537      M  35         14    10003  
1000208     4.0  1001781044      M  18          4    61820  

[1000209 rows x 10 columns]

In [28]:
user_item_matrix = data.pivot(index="UserId", columns="MovieId", values="Rating")

In [29]:
user_item_matrix

MovieId    1  10  100  1000  1002  1003  1004  1005  1006  1007  ...  99  990  \
UserId                                                           ...            
1        5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
10       5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
100      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1000     5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1001     4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
...      ...  ..  ...   ...   ...   ...   ...   ...   ...   ...  ...  ..  ...   
995      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
996      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
997      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
998      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
999      NaN NaN  NaN   NaN   NaN   2.0   3.0   NaN   NaN   NaN  ... NaN  NaN   

MovieId  991  992  993  994  996  997  998  999  
UserId                                           
1        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
10       NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
100      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1000     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1001     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...      ...  ...  ...  ...  ...  ...  ...  ...  
995      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
996      NaN  NaN  NaN  NaN  NaN  NaN  NaN  3.0  
997      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
998      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
999      NaN  NaN  NaN  NaN  2.0  NaN  NaN  NaN  

[6040 rows x 3706 columns]

In [30]:
# for user in data["UserId"].unique():
#     for item in data["MovieId"].unique():
#         if pd.isna(user_item_matrix.loc[user, item]):  # Predict only if NaN
#             predicted_rating = model.predict(uid=user, iid=item).est
#             user_item_matrix.loc[user, item] = round(predicted_rating, 2)  # Store predicted rating

# # Step 7: Print Final Reconstructed Matrix (Original + Predicted)
# print("\nReconstructed User-Item Matrix (Original + Predicted Rating):")
# print(user_item_matrix)

In [31]:
all_user_item_pairs = [
    (uid, iid, 0) for uid in data["UserId"].unique() for iid in data["MovieId"].unique()
]

# Predict in a single batch (faster)
predictions = model.test(all_user_item_pairs)

# Convert predictions to a DataFrame
pred_df = pd.DataFrame([(pred.uid, pred.iid, round(pred.est, 2)) for pred in predictions],
                        columns=["UserId", "MovieId", "PredictedRating"])


In [32]:
all_user_item_pairs

[('1', '1', 0),
 ('1', '2', 0),
 ('1', '3', 0),
 ('1', '4', 0),
 ('1', '5', 0),
 ('1', '6', 0),
 ('1', '7', 0),
 ('1', '8', 0),
 ('1', '9', 0),
 ('1', '10', 0),
 ('1', '11', 0),
 ('1', '12', 0),
 ('1', '13', 0),
 ('1', '14', 0),
 ('1', '15', 0),
 ('1', '16', 0),
 ('1', '17', 0),
 ('1', '18', 0),
 ('1', '19', 0),
 ('1', '20', 0),
 ('1', '21', 0),
 ('1', '22', 0),
 ('1', '23', 0),
 ('1', '24', 0),
 ('1', '25', 0),
 ('1', '26', 0),
 ('1', '27', 0),
 ('1', '28', 0),
 ('1', '29', 0),
 ('1', '30', 0),
 ('1', '31', 0),
 ('1', '32', 0),
 ('1', '33', 0),
 ('1', '34', 0),
 ('1', '35', 0),
 ('1', '36', 0),
 ('1', '37', 0),
 ('1', '38', 0),
 ('1', '39', 0),
 ('1', '40', 0),
 ('1', '41', 0),
 ('1', '42', 0),
 ('1', '43', 0),
 ('1', '44', 0),
 ('1', '45', 0),
 ('1', '46', 0),
 ('1', '47', 0),
 ('1', '48', 0),
 ('1', '49', 0),
 ('1', '50', 0),
 ('1', '52', 0),
 ('1', '53', 0),
 ('1', '54', 0),
 ('1', '55', 0),
 ('1', '56', 0),
 ('1', '57', 0),
 ('1', '58', 0),
 ('1', '59', 0),
 ('1', '60', 0),
 ('1',

In [33]:
predictions

[Prediction(uid='1', iid='1', r_ui=0, est=4.4318409158802625, details={'was_impossible': False}),
 Prediction(uid='1', iid='2', r_ui=0, est=3.5907712181581704, details={'was_impossible': False}),
 Prediction(uid='1', iid='3', r_ui=0, est=3.5316492009679843, details={'was_impossible': False}),
 Prediction(uid='1', iid='4', r_ui=0, est=3.237057725705792, details={'was_impossible': False}),
 Prediction(uid='1', iid='5', r_ui=0, est=3.4432650917043404, details={'was_impossible': False}),
 Prediction(uid='1', iid='6', r_ui=0, est=4.171008640196243, details={'was_impossible': False}),
 Prediction(uid='1', iid='7', r_ui=0, est=3.9421199138936873, details={'was_impossible': False}),
 Prediction(uid='1', iid='8', r_ui=0, est=3.5081658226970527, details={'was_impossible': False}),
 Prediction(uid='1', iid='9', r_ui=0, est=3.2816241686982828, details={'was_impossible': False}),
 Prediction(uid='1', iid='10', r_ui=0, est=3.830983601310575, details={'was_impossible': False}),
 Prediction(uid='1', i

In [34]:
# Convert to user-item matrix
predicted_matrix = pred_df.pivot(index="UserId", columns="MovieId", values="PredictedRating")

print(predicted_matrix)

MovieId     1    10   100  1000  1002  1003  1004  1005  1006  1007  ...  \
UserId                                                               ...   
1        4.43  3.83  3.46  3.37  4.35  3.29  3.23  2.87  3.45  3.37  ...   
10       4.52  4.15  3.68  3.58  4.45  3.52  3.64  3.40  3.73  3.53  ...   
100      3.45  2.88  2.63  2.55  3.42  2.45  2.27  1.86  2.53  2.29  ...   
1000     4.37  3.81  3.37  3.21  4.24  3.19  2.92  2.53  3.30  3.25  ...   
1001     4.23  3.25  3.23  3.19  4.19  3.12  2.50  2.04  3.02  2.95  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
995      4.37  3.74  3.49  3.36  4.40  3.34  3.10  2.76  3.44  3.30  ...   
996      4.51  3.70  3.46  3.36  4.34  3.32  2.91  2.66  3.36  3.31  ...   
997      4.22  3.48  3.27  3.17  4.15  3.15  2.78  2.49  3.21  3.09  ...   
998      4.39  3.55  3.39  3.39  4.30  3.25  3.26  2.89  3.46  3.43  ...   
999      4.07  3.44  2.70  2.39  3.74  2.65  2.22  2.14  2.69  2.74  ...   

MovieId    

In [35]:
from collections import defaultdict

def get_top_n(predictions, n=10):
    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))
    
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)  # Sort by estimated rating
        top_n[uid] = user_ratings[:n]  # Keep top-N

    return top_n

top_n_predictions = get_top_n(predictions, n=5)
print(top_n_predictions)


defaultdict(<class 'list'>, {'1': [('2905', 4.962686679148848), ('527', 4.849006652885293), ('318', 4.7917277818458865), ('668', 4.738171136059803), ('2019', 4.730964300178869)], '6': [('2905', 4.839589406020543), ('905', 4.680233171238649), ('668', 4.658243853793128), ('1035', 4.650886667734769), ('1236', 4.644083060408016)], '8': [('2905', 4.846404448355467), ('858', 4.801295868514962), ('922', 4.791340179363549), ('318', 4.79116742865586), ('593', 4.771676012419379)], '9': [('318', 4.61123410698307), ('593', 4.502292084628309), ('2905', 4.452517666788131), ('50', 4.427278624487316), ('527', 4.418945087169709)], '10': [('2501', 5), ('2905', 5), ('527', 4.986739541006988), ('539', 4.975145300938802), ('668', 4.9695419885766405)], '18': [('912', 5), ('923', 5), ('1148', 5), ('1204', 5), ('1207', 5)], '19': [('2905', 4.574381540895292), ('318', 4.4819917958549595), ('922', 4.429502242967128), ('1236', 4.411288321206268), ('3435', 4.39461187777191)], '21': [('2905', 4.132846761551612), (

In [36]:
from surprise import SVD, Dataset, Reader
from surprise.model_selection import train_test_split
import pandas as pd
import numpy as np
from scipy.sparse import coo_matrix

# ratings_df = Dataset.load_from_df(data[['UserId','MovieId','Rating']], reader)

# Define the dataset for Surprise
reader = Reader(rating_scale=(1, 5))
dataset = Dataset.load_from_df(data[["UserId", "MovieId", "Rating"]], reader)

# Train-test split
trainset, testset = train_test_split(dataset, test_size=0.2)

# Train the SVD model
model = SVD()
model.fit(trainset)

# Create user-item matrix
user_item_matrix = data.pivot(index="UserId", columns="MovieId", values="Rating")

# Convert user-item matrix to a sparse matrix (CSR format)
user_item_matrix_sparse = coo_matrix(user_item_matrix.fillna(0))

# Extract missing values from the sparse matrix
missing_values = np.where(user_item_matrix_sparse.toarray() == 0)
# Ensure that the testset is correctly structured: a list of (user, item, rating) tuples
all_user_item_pairs = [
    (uid, iid, 0) for uid in data["UserId"].unique() for iid in data["MovieId"].unique()
]

# You need to predict on missing values only
predictions = model.test(all_user_item_pairs)


# # Predict ratings for missing values
# predictions = model.test(zip(missing_values[0] + 1, missing_values[1] + 1))  # +1 because Surprise uses 1-based indexing

# Map predictions back to the user-item matrix
predicted_ratings = { (pred.uid, pred.iid): pred.est for pred in predictions }

# Update the original user-item matrix with predicted values for missing ratings
for i, j in zip(missing_values[0], missing_values[1]):
    if (i + 1, j + 1) in predicted_ratings:
        user_item_matrix.iloc[i, j] = round(predicted_ratings[(i + 1, j + 1)], 2)

print(user_item_matrix)


MovieId    1  10  100  1000  1002  1003  1004  1005  1006  1007  ...  99  990  \
UserId                                                           ...            
1        5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
10       5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
100      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1000     5.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
1001     4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
...      ...  ..  ...   ...   ...   ...   ...   ...   ...   ...  ...  ..  ...   
995      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
996      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
997      4.0 NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
998      NaN NaN  NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ... NaN  NaN   
999      NaN NaN  NaN   NaN 

In [37]:
# Loop through missing values and replace them with predicted ratings
for i, j in zip(missing_values[0], missing_values[1]):
    # Get the user and movie IDs (we need to get these from the matrix indices, not integer indices)
    user_id = user_item_matrix.index[i]
    movie_id = user_item_matrix.columns[j]

    # Check if the prediction for this (user, movie) pair exists
    if (user_id, movie_id) in predicted_ratings:
        # Replace the NaN value with the predicted rating
        user_item_matrix.loc[user_id, movie_id] = round(predicted_ratings[(user_id, movie_id)], 2)

print(user_item_matrix)

MovieId     1    10   100  1000  1002  1003  1004  1005  1006  1007  ...  \
UserId                                                               ...   
1        5.00  4.00  3.78  3.84  4.04  3.43  3.19  2.85  3.58  3.44  ...   
10       5.00  3.95  3.67  3.81  4.30  3.51  3.52  3.47  3.81  4.07  ...   
100      3.43  2.93  2.68  2.81  3.20  2.56  2.35  2.04  2.69  2.42  ...   
1000     5.00  3.70  3.41  3.26  4.03  3.34  2.99  3.02  3.50  3.44  ...   
1001     4.00  3.37  3.47  3.50  3.74  3.13  2.36  2.21  3.14  3.12  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
995      4.49  3.59  3.46  3.58  4.07  2.98  2.92  2.42  3.31  3.27  ...   
996      4.00  3.88  3.77  3.76  3.94  3.42  2.87  2.68  3.53  3.22  ...   
997      4.00  3.58  3.17  3.52  4.00  3.19  2.99  2.48  3.28  3.15  ...   
998      4.31  3.76  3.89  3.69  4.40  3.82  3.39  3.30  4.00  3.58  ...   
999      3.88  3.30  3.05  2.93  3.31  2.00  3.00  2.34  2.55  3.05  ...   

MovieId    

In [38]:
from sklearn.metrics.pairwise import cosine_similarity
user_similarity = cosine_similarity(user_item_matrix)
user_similarity

array([[1.        , 0.99381001, 0.99632283, ..., 0.99770536, 0.9969374 ,
        0.98954385],
       [0.99381001, 1.        , 0.99092298, ..., 0.99390221, 0.99474213,
        0.98733352],
       [0.99632283, 0.99092298, 1.        , ..., 0.99662683, 0.99443929,
        0.98660759],
       ...,
       [0.99770536, 0.99390221, 0.99662683, ..., 1.        , 0.9968014 ,
        0.99009874],
       [0.9969374 , 0.99474213, 0.99443929, ..., 0.9968014 , 1.        ,
        0.98774649],
       [0.98954385, 0.98733352, 0.98660759, ..., 0.99009874, 0.98774649,
        1.        ]])

In [39]:
user_similarity_df = pd.DataFrame(user_similarity, index=user_item_matrix.index, columns=user_item_matrix.index)

In [40]:
user_similarity_df

UserId         1        10       100      1000      1001      1002      1003  \
UserId                                                                         
1       1.000000  0.993810  0.996323  0.997814  0.992764  0.996163  0.998155   
10      0.993810  1.000000  0.990923  0.994583  0.983683  0.989319  0.994644   
100     0.996323  0.990923  1.000000  0.996002  0.992888  0.996873  0.996958   
1000    0.997814  0.994583  0.996002  1.000000  0.991311  0.995387  0.998264   
1001    0.992764  0.983683  0.992888  0.991311  1.000000  0.994715  0.992564   
...          ...       ...       ...       ...       ...       ...       ...   
995     0.997477  0.992784  0.996916  0.997412  0.993458  0.997195  0.998127   
996     0.996904  0.990580  0.995818  0.995985  0.992678  0.996286  0.996746   
997     0.997705  0.993902  0.996627  0.998124  0.992047  0.996195  0.998291   
998     0.996937  0.994742  0.994439  0.997384  0.989817  0.994309  0.997428   
999     0.989544  0.987334  0.986608  0.990294  0.979235  0.984296  0.990443   

UserId      1004      1005      1006  ...       990       991       992  \
UserId                                ...                                 
1       0.989583  0.996005  0.994821  ...  0.996922  0.997320  0.997020   
10      0.982891  0.989360  0.986227  ...  0.992570  0.993661  0.991377   
100     0.990127  0.995620  0.995557  ...  0.994941  0.996790  0.996569   
1000    0.990195  0.995146  0.994342  ...  0.996176  0.997463  0.996598   
1001    0.984895  0.991986  0.993454  ...  0.990687  0.991893  0.993125   
...          ...       ...       ...  ...       ...       ...       ...   
995     0.990022  0.995999  0.995955  ...  0.996588  0.997883  0.997299   
996     0.989640  0.996303  0.995699  ...  0.995313  0.996389  0.996285   
997     0.990181  0.996122  0.995491  ...  0.996612  0.998013  0.997458   
998     0.987010  0.993848  0.991899  ...  0.995973  0.996479  0.995425   
999     0.983444  0.986453  0.984683  ...  0.987559  0.988790  0.988036   

UserId       993       994       995       996       997       998       999  
UserId                                                                        
1       0.995186  0.998083  0.997477  0.996904  0.997705  0.996937  0.989544  
10      0.988842  0.994086  0.992784  0.990580  0.993902  0.994742  0.987334  
100     0.995589  0.996343  0.996916  0.995818  0.996627  0.994439  0.986608  
1000    0.995343  0.997867  0.997412  0.995985  0.998124  0.997384  0.990294  
1001    0.992980  0.992653  0.993458  0.992678  0.992047  0.989817  0.979235  
...          ...       ...       ...       ...       ...       ...       ...  
995     0.996508  0.997903  1.000000  0.996546  0.998111  0.996086  0.988503  
996     0.995638  0.996586  0.996546  1.000000  0.996588  0.994883  0.987339  
997     0.996371  0.998239  0.998111  0.996588  1.000000  0.996801  0.990099  
998     0.993645  0.997009  0.996086  0.994883  0.996801  1.000000  0.987746  
999     0.987067  0.989612  0.988503  0.987339  0.990099  0.987746  1.000000  

[6040 rows x 6040 columns]

In [41]:
item_similarity = cosine_similarity(user_item_matrix.T)
item_similarity_df = pd.DataFrame(item_similarity, index=user_item_matrix.columns, columns=user_item_matrix.columns)
item_similarity_df

MovieId         1        10       100      1000      1002      1003      1004  \
MovieId                                                                         
1        1.000000  0.987075  0.989407  0.991684  0.991559  0.989178  0.985008   
10       0.987075  1.000000  0.993188  0.992792  0.993815  0.993615  0.990763   
100      0.989407  0.993188  1.000000  0.996763  0.997261  0.997434  0.994609   
1000     0.991684  0.992792  0.996763  1.000000  0.998264  0.997068  0.993260   
1002     0.991559  0.993815  0.997261  0.998264  1.000000  0.997507  0.994394   
...           ...       ...       ...       ...       ...       ...       ...   
994      0.990188  0.989302  0.993251  0.995807  0.995827  0.993547  0.987542   
996      0.982076  0.990381  0.993585  0.991940  0.993613  0.993877  0.993476   
997      0.991371  0.992811  0.996362  0.997799  0.997970  0.996859  0.992289   
998      0.988865  0.992047  0.996043  0.996826  0.996927  0.996401  0.995884   
999      0.988455  0.991898  0.994431  0.995897  0.996591  0.995277  0.992261   

MovieId      1005      1006      1007  ...        99       990       991  \
MovieId                                ...                                 
1        0.979381  0.988967  0.988843  ...  0.991277  0.988734  0.990280   
10       0.987878  0.993324  0.992554  ...  0.992524  0.992812  0.992635   
100      0.991354  0.997314  0.995023  ...  0.996397  0.996614  0.996501   
1000     0.988502  0.997013  0.994629  ...  0.998055  0.996095  0.996720   
1002     0.990211  0.997918  0.995467  ...  0.997882  0.996909  0.997908   
...           ...       ...       ...  ...       ...       ...       ...   
994      0.981423  0.993424  0.990904  ...  0.996167  0.992285  0.994787   
996      0.990555  0.993979  0.992211  ...  0.991831  0.993721  0.992713   
997      0.987828  0.996602  0.994440  ...  0.998262  0.996050  0.997266   
998      0.992464  0.997175  0.994517  ...  0.995754  0.996918  0.996087   
999      0.988595  0.995849  0.994291  ...  0.995994  0.995148  0.995886   

MovieId       992       993       994       996       997       998       999  
MovieId                                                                        
1        0.988937  0.990768  0.990188  0.982076  0.991371  0.988865  0.988455  
10       0.991906  0.994526  0.989302  0.990381  0.992811  0.992047  0.991898  
100      0.996562  0.997905  0.993251  0.993585  0.996362  0.996043  0.994431  
1000     0.997568  0.997836  0.995807  0.991940  0.997799  0.996826  0.995897  
1002     0.997945  0.998582  0.995827  0.993613  0.997970  0.996927  0.996591  
...           ...       ...       ...       ...       ...       ...       ...  
994      0.994350  0.994632  1.000000  0.987611  0.996552  0.992327  0.993725  
996      0.993693  0.994539  0.987611  1.000000  0.991586  0.993565  0.992119  
997      0.997852  0.998066  0.996552  0.991586  1.000000  0.995337  0.995949  
998      0.996365  0.997090  0.992327  0.993565  0.995337  1.000000  0.995200  
999      0.996115  0.995812  0.993725  0.992119  0.995949  0.995200  1.000000  

[3706 rows x 3706 columns]